plugins/oauth/girder_oauth/providers/microsoft.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+from msal import ConfidentialClientApplication
+
+from girder.exceptions import RestException
+from girder.models.setting import Setting
+
+from ..settings import PluginSettings
+from .base import ProviderBase
+
+
+class Microsoft(ProviderBase):
+    _AUTH_SCOPES = ['User.Read']
+    _API_USER_URL = 'https://graph.microsoft.com/v1.0/me'
+
+    @classmethod
+    def _authority(cls):
+        tenantId = Setting().get(PluginSettings.MICROSOFT_TENANT_ID) or 'common'
+        return f'https://login.microsoftonline.com/{tenantId}'
+
+    def getClientIdSetting(self):
+        return Setting().get(PluginSettings.MICROSOFT_CLIENT_ID)
+
+    def getClientSecretSetting(self):
+        return Setting().get(PluginSettings.MICROSOFT_CLIENT_SECRET)
+
+    @classmethod
+    def getUrl(cls, state):
+        clientId = Setting().get(PluginSettings.MICROSOFT_CLIENT_ID)
+        if not clientId:
+            raise Exception('No Microsoft client ID setting is present.')
+
+        clientSecret = Setting().get(PluginSettings.MICROSOFT_CLIENT_SECRET)
+        if not clientSecret:
+            raise Exception('No Microsoft client secret setting is present.')
+
+        app = ConfidentialClientApplication(
+            client_id=clientId,
+            client_credential=clientSecret,
+            authority=cls._authority(),
+        )
+        # The default response type is 'code', so we don't need to pass it
+        url = app.get_authorization_request_url(
+            scopes=cls._AUTH_SCOPES,
+            state=state,
+        )
+        return url
+
+    def getToken(self, code):
+        app = ConfidentialClientApplication(
+            client_id=self.clientId,
+            client_credential=self.clientSecret,
+            authority=self._authority(),
+        )
+        result = app.acquire_token_by_authorization_code(
```

Comment:
I notice that the msal library says `DeprecationWarning: Change your acquire_token_by_authorization_code() to acquire_token_by_auth_code_flow()`

Reply:
I'll try to use that instead.

Reply:
I think if we use `acquire_token_by_auth_code_flow`, it will take the user to the login page, which is not what we want because that is handled internally by Girder. As this is working as it is, shall I add some code to ignore the deprecation warning instead?

Reply:
Yes -- we can always revisit it in the future if the deprecated flow is actually removed.